File: reactapp/src/types/substitution/conceptual-knowledge/substitutionExercise.ts
import { Type } from "class-transformer";
import {
    ISubstitutionVariable,
    SubstitutionVariable
} from "@/types/substitution/conceptual-knowledge/substitutionVariable.ts";
import {
    ISubstitutionEquation,
    SubstitutionEquation
} from "@/types/substitution/conceptual-knowledge/substitutionEquation.ts";

export interface ISubstitutionExercise {
    readonly id: number;
    readonly firstEquation: ISubstitutionEquation;
    readonly secondEquation: ISubstitutionEquation;
    readonly isolatedVariable: ISubstitutionVariable;
    readonly secondVariable: ISubstitutionVariable;
}

export class SubstitutionExercise implements ISubstitutionExercise {
    public readonly id: number;

    @Type(() => SubstitutionEquation)
    public readonly firstEquation: SubstitutionEquation;

    @Type(() => SubstitutionEquation)
    public readonly secondEquation: SubstitutionEquation;

    @Type(() => SubstitutionVariable)
    public readonly isolatedVariable: SubstitutionVariable;

    @Type(() => SubstitutionVariable)
    public readonly secondVariable: SubstitutionVariable;

    constructor(
        id: number,
        firstEquation: SubstitutionEquation,
        secondEquation: SubstitutionEquation,
        isolatedVariable: SubstitutionVariable,
        secondVariable: SubstitutionVariable
    ) {
        this.id = id;
        this.firstEquation = firstEquation;
        this.secondEquation = secondEquation;
        this.isolatedVariable = isolatedVariable;
        this.secondVariable = secondVariable;
    }
}

File: webapi/Models/Substitution/ConceptualKnowledge/SubstitutionExercise.cs
using webapi.Models.ConceptualKnowledge;

namespace webapi.Models.Substitution.ConceptualKnowledge
{
    public class SubstitutionExercise : ICKExercise
    {
        public long? Id { get; set; }

        public int? Ordering { get; set; }

        public int? Level { get; set; }

        public required SubstitutionEquation FirstEquation { get; set; }

        public required SubstitutionEquation SecondEquation { get; set; }

        public required SubstitutionVariable IsolatedVariable { get; set; }

        public required SubstitutionVariable SecondVariable { get; set; }
    }
}
